omc3/kmod/analysis.py

```diff
@@ -247,41 +254,147 @@ def return_df(magnet1_df, magnet2_df, plane):
     elif check_polarity(magnet1_df, magnet2_df, -sign[plane]):
         return magnet2_df, magnet1_df
 
+      
+def get_BPM(kmod_input_params):
+
+    # listing the BPMs of the last quadrupole BPMs
+    BPM_dict = {
+        "IP1" : ["BPMSW.1L1", "BPMSW.1R1"], 
+        "IP2" : ["BPMSW.1L2", "BPMSW.1R2"], 
+        "IP3" : ["BPMW.4L3", "BPMW.4R3"], 
+        "IP4" : ["BPMWA.A5L4", "BPMWA.A5R4"], 
+        "IP5" : ["BPMSW.1L5", "BPMSW.1R5"],
+        "IP6" : ["BPMSE.4L6", "BPMSA.4R6"], 
+        "IP7" : ["BPMW.4L7", "BPMW.4R7"], 
+        "IP8" : ["BPMSW.1L8", "BPMSW.1R8"], 
+    }
+
+    # Selecting BPMs
+    BPML = BPM_dict[kmod_input_params.ip.upper()][0] + '.' + kmod_input_params.beam.upper()
+    BPMR = BPM_dict[kmod_input_params.ip.upper()][1] + '.' + kmod_input_params.beam.upper()
+
+    return BPML, BPMR
+
+def get_BPM_distance(kmod_input_params,BPML,BPMR):
+
+    model_filename = 'twiss_' + kmod_input_params.beam.lower() + '.dat'
+
+    # Position s of BPML
+    if (os.path.exists(os.path.join(f'{kmod_input_params.twiss_model_dir}', f'{model_filename}'))): # distance between actual BPMs if model is provided (more precise)
+        twiss_df = tfs.read(os.path.join(f'{kmod_input_params.twiss_model_dir}', f'{model_filename}'), index='NAME')
+        pos_1L = twiss_df.loc[BPML,'S']
+        pos_1R = twiss_df.loc[BPMR,'S']
+        BPM_distance = np.abs(pos_1R - pos_1L)/2.0
+    else:
+        BPM_distance = 22.965 # Distance between last quadrupole and IP (LSTAR) (less precise)
```

Comment:
is this the same for all IPs? if so leave just a number, if not not model is required or there are also sequence files...
Do we need the model at all?

Reply:
Eventually the modulation could be performed in Q2 so the distance between BPMs would change. In this case, the BPM names in the dict above must be changed as well. Need to think how to implement this. In any case, using the sequence and not the model is better.